src/lib.rs

```diff
@@ -251,110 +247,116 @@ impl<'a, T> core::clone::Clone for Menu<'a, T> {
+                write!(context, "/").unwrap();
             }
 
             let menu = self.menu_mgr.get_menu(Some(i));
-            write!(self.context, "/{}", menu.label).unwrap();
+            write!(context, "/{}", menu.label).unwrap();
         }
-        write!(self.context, "> ").unwrap();
+        write!(context, "> ").unwrap();
     }
 
-    /// Add a byte to the menu runner's buffer. If this byte is a
-    /// carriage-return, the buffer is scanned and the appropriate action
-    /// performed.
+    /// Process input data for command lines.
+    ///
     /// By default, an echo feature is enabled to display commands on the terminal.
-    pub fn input_byte(&mut self, input: u8) {
-        // Strip carriage returns
-        if input == 0x0A {
-            return;
-        }
-        let outcome = if input == 0x0D {
-            #[cfg(not(feature = "echo"))]
-            {
-                // Echo the command
-                write!(self.context, "\r").unwrap();
-                if let Ok(s) = core::str::from_utf8(&self.buffer[0..self.used]) {
-                    write!(self.context, "{}", s).unwrap();
-                }
-            }
-            // Handle the command
-            self.process_command();
-            Outcome::CommandProcessed
-        } else if (input == 0x08) || (input == 0x7F) {
-            // Handling backspace or delete
-            if self.used > 0 {
-                write!(self.context, "\u{0008} \u{0008}").unwrap();
-                self.used -= 1;
+    pub fn process(&mut self, context: &mut T) {
+        while context.read_ready().unwrap() {
+            let mut input_buf = [0; 1];
+            context.read(&mut input_buf).unwrap();
+            let input = input_buf[0];
+
+            // Strip carriage returns
+            if input == 0x0A {
+                continue;
             }
-            Outcome::NeedMore
-        } else if self.used < self.buffer.len() {
-            self.buffer[self.used] = input;
-            self.used += 1;
 
-            #[cfg(feature = "echo")]
-            {
-                // We have to do this song and dance because `self.prompt()` needs
-                // a mutable reference to self, and we can't have that while
-                // holding a reference to the buffer at the same time.
-                // This line grabs the buffer, checks it's OK, then releases it again
-                let valid = core::str::from_utf8(&self.buffer[0..self.used]).is_ok();
-                // Now we've released the buffer, we can draw the prompt
-                if valid {
-                    write!(self.context, "\r").unwrap();
-                    self.prompt(false);
+            let outcome = if input == 0x0D {
+                #[cfg(not(feature = "echo"))]
+                {
+                    // Echo the command
+                    write!(context, "\r").unwrap();
+                    if let Ok(s) = core::str::from_utf8(&self.buffer[0..self.used]) {
```

Comment:
This doesn't need stringifying, right? We can just throw the bytes using the embedded-io::Write that T impls?

Reply:
Fixed in a8159862066c78bac980c7a302f26f04a86d1554